src/lib/ledger_catchup/ledger_catchup.ml

```diff
@@ -254,32 +146,33 @@ module Make (Inputs : Inputs.S) :
                       peer ;
                     Deferred.return (Error (Error.of_string "")) )
                 in
-                let%bind verified_transitions =
-                  take_while_map_result_rev rest
+                let%bind verified_transitions, initial_state_hash =
+                  take_while_map_result_rev
+                    Non_empty_list.(to_list rev_queried_transitions)
                     ~f:
                       (verify_transition ~logger ~frontier
                          ~unprocessed_transition_cache)
                 in
-                let%bind () =
-                  Deferred.return
-                    ( if List.length verified_transitions > 0 then Ok ()
-                    else
-                      let error =
-                        "Peer should have given us some new transitions that \
-                         are not in our transition frontier"
-                      in
-                      Logger.faulty_peer logger ~module_:__MODULE__
-                        ~location:__LOC__ "%s" error ;
-                      Error (Error.of_string error) )
-                in
-                let full_subtree =
-                  List.fold_right verified_transitions ~init:target_subtree
-                    ~f:(fun transition acc -> Rose_tree.T (transition, [acc])
-                  )
+                let subtrees_of_transitions =
+                  if List.length verified_transitions > 0 then
+                    let rest, target_transition_singleton =
+                      List.split_n verified_transitions
```

Comment:
The creation of `subtrees_of_transitions` is hard to read, namely because of `split_n`. Can you rewrite this without using `split_n` and `hd_exn`?